### PR TITLE
CrystFEL: support aarch64-linux and HDF5 on Darwin

### PR DESCRIPTION
**Pull request by pmiddend**
###### Description of changes

The CrystFEL software used to not run on aarch64. This is due to:

1. An old version of the dependent library libccp4, which couldn't find the `gfortran` dependency. There is a newer version, 8.0.0 of this software, but this didn't work out of the box either (the CMake build process assumes certain unnecessary libraries to be installed), so the author of CrystFEL recommended to me to use a patch that adds the meson build systems instead. This is also used by the official package, see [here](https://gitlab.desy.de/thomas.white/crystfel/-/blob/doc-test/subprojects/libccp4c.wrap) (in nixpkgs, I avoided the wrap functionality and packaged the dependencies using Nix instead). It does not depend on gfortran anymore, it only builds the C version (for CrystFEL, only that is important).
2. On `aarch64-darwin`, the function `fmemopen` doesn't exist (used to map a region of memory to a `FILE*` pointer). The corresponding code I patched and sent the patch upstream (see [here](https://gitlab.desy.de/thomas.white/crystfel/-/merge_requests/10)). On `aarch64-linux`, we have the `memorymappingHook` which (still) works.

Also, the HDF5-External-Filter-Plugin didn't work on Darwin because of a linker issue. This has been fixed by an external contributor.

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [x] aarch64-linux
  - [x] x86_64-darwin
  - [x] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.05 Release Notes (or backporting 22.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2305-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by pmiddend** [comment]
@ofborg build crystfel-headless

**Comment by pmiddend** [comment]
@ofbord build crystfel

File: pkgs/applications/science/physics/crystfel/default.nix [hunk 1]
```diff
@@ -38,30 +37,25 @@
...
   libccp4 = stdenv.mkDerivation rec {
     pname = "libccp4";
-    version = "6.5.1";
+    version = "8.0.0";
     src = fetchurl {
-      # Original mirror, now times out
-      # url = "ftp://ftp.ccp4.ac.uk/opensource/${pname}-${version}.tar.gz";
-      url = "https://deb.debian.org/debian/pool/main/libc/${pname}/${pname}_${version}.orig.tar.gz";
-      sha256 = "1rfvjliny29vy5bdi6rrjaw9hhhhh72pw536xwvqipqcjlylf2r8";
+      url = "http://ftp.ccp4.ac.uk/opensource/${pname}-${version}.tar.gz";
```
Comment (SuperSandro2000) [inline pkgs/applications/science/physics/crystfel/default.nix:42]:
```suggestion
      url = "https://ftp.ccp4.ac.uk/opensource/${pname}-${version}.tar.gz";
```

File: pkgs/applications/science/physics/crystfel/default.nix [hunk 2]
```diff
@@ -38,30 +37,25 @@
...
-    preBuild = ''
-      makeFlagsArray+=(CFLAGS='-DNIX_PROVIDED_SYMOP_FILE=\"${placeholder "out"}/share/syminfo.lib\"')
-      export NIX_LDFLAGS="-L${gfortran.cc}/lib64 -L${gfortran.cc}/lib $NIX_LDFLAGS";
-    '';
-    makeFlags = [ "CFLAGS='-DNIX_PROVIDED_SYMOP_FILE=\"${placeholder "out"}/share/syminfo.lib\"" ];
+    env.NIX_CFLAGS_COMPILE = "-DNIX_PROVIDED_SYMOP_FILE=\"${placeholder "out"}/share/ccp4/syminfo.lib\"";
 
     patches = [
+      # This circumvents the original autoconf/CMake based build and uses meson instead
+      ./add-meson-build.patch
```
Comment (SuperSandro2000) [inline pkgs/applications/science/physics/crystfel/default.nix:57]: Did you write that yourself? If so did you try to upstream it?

File: pkgs/applications/science/physics/crystfel/default.nix [hunk 3]
```diff
@@ -143,10 +137,10 @@ let
     pname = "HDF5-External-Filter-Plugins";
     version = "0.1.0";
     src = fetchFromGitHub {
-      owner = "nexusformat";
+      owner = "spanezz";
```
Comment (SuperSandro2000) [inline pkgs/applications/science/physics/crystfel/default.nix:140]: The only difference is one commit https://github.com/nexusformat/HDF5-External-Filter-Plugins/compare/master...spanezz:HDF5-External-Filter-Plugins:master so we can fetchpatch that

File: pkgs/applications/science/physics/crystfel/default.nix [hunk 4]
```diff
@@ -224,7 +219,7 @@ stdenv.mkDerivation rec {
     downloadPage = "https://www.desy.de/~twhite/crystfel/download.html";
     license = licenses.gpl3Plus;
     maintainers = with maintainers; [ pmiddend ];
-    platforms = [ "x86_64-linux" "x86_64-darwin" ];
+    platforms = [ "x86_64-linux" "x86_64-darwin" "aarch64-linux" "aarch64-darwin" ];
```
Comment (SuperSandro2000) [inline pkgs/applications/science/physics/crystfel/default.nix:222]: Maybe just set unix?

File: pkgs/applications/science/physics/crystfel/add-meson-build.patch [hunk 5]
```diff
@@ -0,0 +1,116 @@
+diff --git a/meson.build b/meson.build
+new file mode 100644
+index 0000000..8fc0631
+--- /dev/null
++++ b/meson.build
+@@ -0,0 +1,110 @@
++# Meson file for libccp4c (CCP4 core libraries, C part only)
```
Comment (SuperSandro2000) [inline pkgs/applications/science/physics/crystfel/add-meson-build.patch:7]: If we add the entire file we can skip the patch part which adds some nice highlighting

File: pkgs/applications/science/physics/crystfel/default.nix (hunk 1 above)
Comment (pmiddend) [reply pkgs/applications/science/physics/crystfel/default.nix]: Fixed, thanks!

File: pkgs/applications/science/physics/crystfel/default.nix (hunk 3 above)
Comment (pmiddend) [reply pkgs/applications/science/physics/crystfel/default.nix]: See comment below.

**Comment by pmiddend** [comment]
@SuperSandro2000 Thanks for the review and for pressing me on the meson patch thingy. The patch is actually in meson's "wrapdb" from where I shamelessly copied it.

Now it's downloaded from there. I'm not sure about the new `postUnpack` attribute though. The problem was that wrapdb doesn't give you a patch file, but a zip archive wi `libccp4-8.0.0/meson.build` in it, so I opted for the `fetchzip` + `cp` option. Does that seem fine with you?

File: pkgs/applications/science/physics/crystfel/default.nix (hunk 4 above)
Comment (pmiddend) [reply pkgs/applications/science/physics/crystfel/default.nix]: Yeah, seems to be the same, thanks for the hint.

File: pkgs/applications/science/physics/crystfel/default.nix [hunk 6]
```diff
@@ -145,10 +149,17 @@ let
     src = fetchFromGitHub {
       owner = "nexusformat";
       repo = pname;
-      rev = "d469f175e5273c1d488e71a6134f84088f57d39c";
-      sha256 = "1jrzzh75i68ad1yrim7s1nx9wy0s49ghkziahs71mm5azprm6gh9";
+      rev = "master";
```
Comment (SuperSandro2000) [inline pkgs/applications/science/physics/crystfel/default.nix:152]: This is not reproducible. You need to use the hash of the current head commit of the branch

**Review by SuperSandro2000 — CHANGES_REQUESTED** [review]

File: pkgs/applications/science/physics/crystfel/default.nix (hunk 6 above)
Comment (pmiddend) [reply pkgs/applications/science/physics/crystfel/default.nix]: You're right, of course. Fixed.

**Comment by pmiddend** [comment]
@SuperSandro2000 Ping

**Comment by wegank** [comment]
@ofborg build crystfel crystfel.passthru.tests

**Comment by wegank** [comment]
Result of `nixpkgs-review pr 218968` run on aarch64-darwin [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>2 packages built:</summary>
  <ul>
    <li>crystfel</li>
    <li>crystfel-headless</li>
  </ul>
</details>

**Pull request merged**